### PR TITLE
Add new `ape-pie` client

### PR DESCRIPTION
**Pull request by Viicos**
Should this be the last version in the `0.x` series? This way we will have the added deprecation warning for at least one version, and we can safely remove the old client in version `1.x`.

~~Two [test](https://github.com/open-formulieren/open-forms/blob/master/src/zgw_consumers_ext/tests/test_client_factory.py) [files](https://github.com/open-formulieren/open-forms/blob/master/src/zgw_consumers_ext/tests/test_nlx_rewriting.py) are still to be added there, but they use unittest and not pytest like the rest of the tests. How should I add them here?~~ done

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (sergei-maertens) [inline .github/workflows/ci.yml:24]: django 3.2 supports py310 as maximum version, so you'll need to exclude 311 and 312 from the build matrix.

I don't think that any Django version supports python 3.12 since it was only released two weeks ago, I suspect 4.2 will gain support and 5.0, but older versions seem unlikely. Confirmed by https://code.djangoproject.com/ticket/34118

File: .github/workflows/ci.yml
Comment (sergei-maertens) [inline .github/workflows/ci.yml:84]: better to follow the official recommendation of pinning to `@release/v1`

Comment (Viicos) [reply .github/workflows/ci.yml:84]:
```suggestion
        uses: pypa/gh-action-pypi-publish@v1
```

File: README.rst
Comment (sergei-maertens) [inline README.rst:11]: I'm in favour of removing this alltogether - the `|python-versions|` badge already reflects this

File: setup.cfg
Comment (sergei-maertens) [inline setup.cfg:30]: let's hold off on that

File: setup.cfg
Comment (sergei-maertens) [inline setup.cfg:49]: wouldn't put this in the base requirements, instead add an extra for test utilities and lump it in there so we can do something like `pip install zgw-consumers[testutils]`

File: zgw_consumers/client.py
Comment (sergei-maertens) [inline zgw_consumers/client.py:13]: `get_client_class` and `load_schema_file` are not documented as public API on https://zgw-consumers.readthedocs.io/en/latest/, so we might even get away with not exposing these here.

Looking at https://zgw-consumers.readthedocs.io/en/latest/quickstart.html#usage, we might even skip the compatibility layer alltogether, as only the `Service` model is really considered public API. Of course, the changelog can mention that these names are moved to `legacy` and tell you which imports to update to accomodate users.

File: zgw_consumers/client.py
Comment (sergei-maertens) [inline zgw_consumers/client.py:19]: Can we make the bound `APIClient` rather than `NLXClient`? I think in theory users can use this `build_client` with a different base class and skip the NLX mechanics.

File: zgw_consumers/client.py
Comment (sergei-maertens) [inline zgw_consumers/client.py:45]: this was the method that had the `case` / `match`, right? Can you reinstate that? I'm anticipating future other auth mechanisms, like Oauth2.

File: zgw_consumers/legacy/client.py
Comment (sergei-maertens) [inline zgw_consumers/legacy/client.py:17]: if we're going towards 1.0, let's drop support for zds client < 2.0 too.

Probably best to create a separate PR for that.

File: zgw_consumers/models/services.py
Comment (sergei-maertens) [inline zgw_consumers/models/services.py:207]: you can use `from __future__ import annotations` so you don't have to quote the thing, and then you can do:

```suggestion
    def get_client(cls, url: str, **kwargs) -> ZGWClient | None:
```

File: zgw_consumers/models/services.py
Comment (sergei-maertens) [inline zgw_consumers/models/services.py:25]: this is odd - it worked before moving this to the legacy module, why does it break now?

File: zgw_consumers/test/factories.py
Comment (sergei-maertens) [inline zgw_consumers/test/factories.py:20]: this one needs fixing before we merge this - api_root is URL validated so it must have a protocol and netloc

File: zgw_consumers/test/factories.py
Comment (sergei-maertens) [inline zgw_consumers/test/factories.py:29]: this also depends on a new simple-cert manager version that ships the factories before we can merge this, and then the minimum version of simple-certmanager will need to be set in `setup.cfg`

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

File: zgw_consumers/test/factories.py
Comment (Viicos) [reply zgw_consumers/test/factories.py:29]: Yes just saw that when trying to add tests from open-forms, will first tackle simple cert manager

File: zgw_consumers/legacy/client.py
Comment (Viicos) [reply zgw_consumers/legacy/client.py:17]: Will do in a separate PR

File: zgw_consumers/client.py
Comment (Viicos) [reply zgw_consumers/client.py:19]: I was wondering the same

File: setup.cfg
Comment (Viicos) [reply setup.cfg:49]: Ok, then I'll do the same for `Faker` which is also in the "main" section

File: zgw_consumers/models/services.py
Comment (Viicos) [reply zgw_consumers/models/services.py:25]: The new stuff in `zgw_consumers.client` is itself importing `Service`, but that might not be necessary if I go with https://github.com/maykinmedia/zgw-consumers/pull/73#discussion_r1350292086

File: zgw_consumers/test/factories.py
Comment (Viicos) [reply zgw_consumers/test/factories.py]: https://github.com/maykinmedia/zgw-consumers/pull/73/commits/b5114610c57e64e8b0f6f6e9b20ff5a9d3a8e12c
Is this what you expected?

File: .github/workflows/ci.yml
Comment (sergei-maertens) [inline .github/workflows/ci.yml:84]: the `v1` ref does not exist

```suggestion
        uses: pypa/gh-action-pypi-publish@release/v1
```

File: CHANGELOG.rst
Comment (sergei-maertens) [inline CHANGELOG.rst:11]:
```suggestion
* ``zgw_consumers`` now recommends using ``ape-pie`` as an HTTP client. The old ``ZGWClient``
  is still made available under the ``zgw_consumers.legacy`` module, it is planned to be removed
  in the next major version.
```

File: docs/testing.rst
Comment (sergei-maertens) [inline docs/testing.rst:5]:
```suggestion
   The ``zgw_consumers.test`` module requires some additional dependencies. The easiest way to get them is to
   install the `testutils` extra:
   
   .. code-block:: bash
   
       pip install zgw_consumers[testutils]
```

File: docs/testing.rst
Comment (sergei-maertens) [inline docs/testing.rst:10]:
```suggestion

.. automodule:: zgw_consumers.test.factories
   :members:
```

Let's just use autodoc to document all the available factories :)

File: zgw_consumers/test/factories.py
Comment (sergei-maertens) [inline zgw_consumers/test/factories.py:8]: I think the problem here is that no path is generated in the URL and there is no guarantee that the path ends with a trailing slash - this is normalized in the model at `save` time, but that obviously doesn't get called with `ServiceFactory.build()`, so the factory may generate not-representative api_root values.

I think we should bite the bullet and implement a custom faker provider that inherits from the existing one, using the `url` & generate a valid URL path with trailing slash.

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

File: zgw_consumers/test/factories.py
Comment (Viicos) [reply zgw_consumers/test/factories.py]: https://github.com/maykinmedia/zgw-consumers/pull/73/commits/242477896f09c435a87e905074ae92defeb94290

**Pull request merged**
